src/SBSpergel.cpp

```diff
@@ -0,0 +1,524 @@
+            // approximate 1+k^2 ~ k^2 => good enough:
+            _maxk = std::pow(_gsparams->maxk_threshold, -1./(2*(1+_nu)));
+        }
+        return _maxk;
+    }
+
+    double SpergelInfo::getHLR() const
+    {
+        if (_re == 0.0) _re = calculateFluxRadius(0.5);
+        return _re;
+    }
+
+    double SpergelInfo::getXNorm() const
+    { return std::pow(2., -_nu) / _gamma_nup1 / (2.0 * M_PI); }
+
+    double SpergelInfo::xValue(double r) const
+    {
+        if (r == 0.) return _xnorm0;
+        else return boost::math::cyl_bessel_k(_nu, r) * std::pow(r, _nu);
+    }
+
+    double SpergelInfo::kValue(double ksq) const
+    {
+        return std::pow(1. + ksq, -1. - _nu);
+    }
+
+    class SpergelNuPositiveRadialFunction: public FluxDensity
+    {
+    public:
+        SpergelNuPositiveRadialFunction(double nu, double xnorm0):
+            _nu(nu), _xnorm0(xnorm0) {}
+        double operator()(double r) const {
+            if (r == 0.) return _xnorm0;
+            else return boost::math::cyl_bessel_k(_nu, r) * std::pow(r,_nu);
+        }
+    private:
+        double _nu;
+        double _xnorm0;
+    };
+
+    class SpergelNuNegativeRadialFunction: public FluxDensity
+    {
+    public:
+        SpergelNuNegativeRadialFunction(double nu, double rmin, double a, double b):
+            _nu(nu), _rmin(rmin), _a(a), _b(b) {}
+        double operator()(double r) const {
+            if (r <= _rmin) {
+                return _a + _b*r;
+            } else return boost::math::cyl_bessel_k(_nu, r) * std::pow(r,_nu);
+        }
+    private:
+        double _nu;
+        double _rmin;
+        double _a;
+        double _b;
+    };
+
+    boost::shared_ptr<PhotonArray> SpergelInfo::shoot(int N, UniformDeviate ud) const
+    {
+        dbg<<"SpergelInfo shoot: N = "<<N<<std::endl;
+        dbg<<"Target flux = 1.0\n";
+
+        if (!_sampler) {
+            // Set up the classes for photon shooting
+            double shoot_rmax = calculateFluxRadius(1. - _gsparams->shoot_accuracy);
+            if (_nu > 0.) {
+                std::vector<double> range(2,0.);
+                range[1] = shoot_rmax;
+                _radial.reset(new SpergelNuPositiveRadialFunction(_nu, _xnorm0));
+                _sampler.reset(new OneDimensionalDeviate( *_radial, range, true, _gsparams));
+            } else {
+                // exact s.b. profile diverges at origin, so replace the inner most circle
+                // (defined such that enclosed flux is shoot_acccuracy) with a linear function
+                // that contains the same flux and has the right value at r = rmin.
+                // So need to solve the following for a and b:
+                // int(2 pi r (a + b r) dr, 0..rmin) = shoot_accuracy
+                // a + b rmin = K_nu(rmin) * rmin^nu
+                double flux_target = _gsparams->shoot_accuracy;
+                double shoot_rmin = calculateFluxRadius(flux_target);
```

Comment:
I worked ab it on my idea of making this radius be where the integrated *error* in the flux inside r=R is shoot_accuracy rather than the integrated flux, using the small r approximation of K_niu(r).  

I was hoping there was an analytic solution (using the approximate form for K_nu at least), which would obviate the need for calling calculateFluxRadius with the really small argument for the flux_target, but I kind of ran into a wall when I got to figuring out where the error flipped sign.   The zero crossing seems to require a non-linear root solver, so it would probably be a bit complicated to get this idea working.

So I tried out a few concrete values of nu to see if this idea was worth pursuing.  i.e. How conservative is this choice of rmin?

It turns out that for nu only mildly less than 0, this choice of rmin is a bit conservative.  At nu=-0.3, it is about a factor of 5 smaller than it needs to be.  But as nu gets very small, say near -0.8, the difference gets less important.  It comes out almost the same either way.  And since these are the values where the numerical problems were happening, it seems that there isn't really any benefit to this idea.

So good that you figured out the XTolerance thing to make this work as is.  :)